pkg/tlsx/openssl/openssl.go

```diff
@@ -148,49 +105,78 @@ func (c *Client) ConnectWithOptions(hostname, ip, port string, options clients.C
 		IP:                  resolvedIP,
 		ProbeStatus:         true,
 		Port:                port,
-		Cipher:              tlsCipher,
+		Version:             resp.Session.getTLSVersion(),
+		CertificateResponse: c.convertCertificateToResponse(hostname, resp.AllCerts[0]),
+		Cipher:              resp.Session.Cipher,
 		TLSConnection:       "openssl",
-		CertificateResponse: ztls.ConvertCertificateToResponse(c.options, hostname, x509LeafCertificate),
-		ServerName:          serverName,
+		ServerName:          opensslOptions.ServerName,
 	}
+
+	// Note: openssl s_client does not return server certificate if certificate chain is requested
 	if c.options.TLSChain {
-		for _, opensslCert := range certificateChain {
-			x509Cert, err := c.convertOpenSSLToX509Certificate(opensslCert)
-			if err != nil {
-				return nil, errors.Wrap(err, "could not convert openssl chain certificate")
-			}
-			response.Chain = append(response.Chain, ztls.ConvertCertificateToResponse(c.options, hostname, x509Cert))
+		responses := []*clients.CertificateResponse{}
+		certs := getCertChain(ctx, opensslOptions)
+		for _, v := range certs {
+			responses = append(responses, c.convertCertificateToResponse(hostname, v))
 		}
+		response.Chain = responses
 	}
 	return response, nil
 }
 
-func (c *Client) convertOpenSSLToX509Certificate(opensslCert *openssl.Certificate) (*x509.Certificate, error) {
-	pemBytes, err := opensslCert.MarshalPEM()
-	if err != nil {
-		return nil, errors.Wrap(err, "could not marshal openssl to pem x509")
-	}
-	pemBlock, _ := pem.Decode(pemBytes)
-	if err != nil {
-		return nil, errors.Wrap(err, "could not read openssl pem x509 to go pem")
-	}
-	if pemBlock.Type != "CERTIFICATE" {
-		return nil, errors.Wrap(err, "unsupported pem block type")
-	}
-	x509Certificate, err := x509.ParseCertificate(pemBlock.Bytes)
-	if err != nil {
-		return nil, errors.Wrap(err, "could not convert openssl x509 to go x509")
-	}
-
-	return x509Certificate, nil
+// same as tls
+func (c *Client) convertCertificateToResponse(hostname string, cert *x509.Certificate) *clients.CertificateResponse {
```

Comment:
Again thinking out loud, If the logic is repeated, there is a chance to refactor this method in a common place and avoid redundancies?